From d3f47ab343f9c5e84947cc2447016743839ffa81 Mon Sep 17 00:00:00 2001
From: Carlos Leija <cileija@ti.com>
Date: Thu, 10 Mar 2011 09:37:43 -0600
Subject: OMAP4: Adding check in case CPU1 wakes up before CPU0 from device off or MPU OFF

It is imperative to enforce CPU0 to be ON before CPU1 can call any PPA HAL or
ROM HAL apis. Current handling of this scenario in kernel is not enough.
If CPU1 wakes up before CPU0, it should go to OFF again and await for CPU0 to
awaken him.

Change-Id: If1bece8ca8dff6664e80fa3dec4ba763d8fcc0ad
Signed-off-by: Carlos Leija <cileija@ti.com>
Signed-off-by: Dan Murphy <dmurphy@ti.com>
---
 arch/arm/mach-omap2/sleep44xx.S |   26 +++++++++++++++++++++++---
 1 files changed, 23 insertions(+), 3 deletions(-)

Index: kernel/arch/arm/mach-omap2/sleep44xx.S
===================================================================
--- kernel.orig/arch/arm/mach-omap2/sleep44xx.S	2011-04-02 16:18:59.000000000 +0900
+++ kernel/arch/arm/mach-omap2/sleep44xx.S	2011-04-02 16:55:38.000000000 +0900
@@ -79,6 +79,28 @@
 	beq	do_WFI
 	bne	context_save		@ Save context if needed
 
+	/*	 *     Check the WakeUp CPU
+	 * CPU1 must check if CPU0 is alive/awaken;
+	 * if PL310 is OFF, MPUSS was OFF and CPU0 is still off,
+	 * CPU1 must go to sleep and wait for CPU0
+	 */
+	ldr	r2, =OMAP44XX_L2CACHE_BASE
+	ldr	r0, [r2, #L2X0_CTRL]
+	and	r0, #0x0f
+	cmp	r0, #1
+	beq	set_ns_smp_access
+	mov     r0, #0x03		@ CPU1 to OFF state
+	mov     r1, #0x00		@ Secure L1 is already clean
+	ldr	r12, =0x108		@ SCU power state secure API
+	dsb
+	smc	#0
+	dsb				@ Necessary barriers before wfi
+	dmb
+	isb
+	wfi				@ wait for interrupt
+	nop
+	nop
+	
 restore_context:
         /*
 	 * Check  the wakeup CPU
